### ISSUE TITLE
Training losses going to NaN

### ISSUE THREAD

**Issue by SamuelMidgley** [pz-white/COM6911-DTI-2021#9]
Hi @pz-white,
We have now got a drug similarity matrix and are using this in the code sent previously. We are encountering an issue with training where no matter the parameter combination the training loss and test loss start increasing and quickly go to NaN at around the same point. The data is in logscale but you had previously mentioned using batch normalisation and or activation functions to prevent this. I understand how this would work with a Neural Network but I'm not sure how to implement this in Matrix Factorisation.

![image](https://user-images.githubusercontent.com/80775550/117689645-e12bbd80-b1b1-11eb-98c5-c2c8921ac5a8.png)

**Comment by pz-white** [comment]
Hi @SamuelMidgley 
Do you mean this [PR](https://github.com/pz-white/COM6911-DTI-2021/pull/10)? It has some conflicts and cannot be merged. Could I think your MSCMF code is in the branch Delta-UploadAll: BaseMF with Drug Similarity.ipynb? Thanks.

**Comment by pz-white** [comment]
It should be the learning rate issue and please set a lower one. Actually it is pretty normal with a low learning rate especially for a MF-based model training.

**Comment by pz-white** [comment]
In addtion, could you provide the file **drug_sim.txt** as I cannot successfully run your code **drug_sim.py** to generate it. I'm not sure whether the similarity is normalized, if not, I suggest to normalize the matrix.

**Comment by pz-white** [comment]
Could you explain why the user similarity is embbed and initialized as learned parameters? Could the target similarity be calculated on the smallest dataset Kd?

        self.user_sim = torch.nn.Embedding(n_users, 1)
        torch.nn.init.xavier_uniform_(self.user_sim.weight)

**Comment by SamuelMidgley** [comment]
Hi @pz-white 

It is now running with a learning rate of 0.005, I tried this before but after 300 epochs it had only just crossed an error value of 2.0 for the test set so I am now trying with 600 epochs. 

Yes sorry I completely forgot to do this it is now uploaded onto the google drive, link is [HERE](https://drive.google.com/file/d/1RLJoOzbQFuQhjNpSfI-QL7jaE6pOB_jJ/view?usp=sharing).  

>  I suggest to normalize the matrix - By normalised do you mean the sum of the entries per row must be one? I tried this and it was giving a slightly higher loss for each epoch at least for the first 100 epochs. 

That was my solution to dealing with only giving the model a batch, The embedding matrix is a weight matrix (separate to the similarity matrix) to assign each drug a weight so that my additional code in the forward function works and has the correct dimensions. Now looking at it again though I'm not sure if this is correct implementation. I did not want to do the full `S = AA^T` for each batch as this would be computationally expensive and also adjust weights for drugs that did not appear in the batch, however, is that the intended method.

The target similarity matrix for Kd dataset is currently running on HPC but has taken upwards of 24 hours now with 30 cores and is soon to run out of allocated time, it has also not even completed 20% of it which leads me to believe the code is either very inefficient or there is an error that it is struggling with so I am currently trying to debug. 

On a similar note I know it is not ideal but time is becoming short, would it be okay to only use say 20-30% of the data for IC50 and Ki as the drug similarity matrices take a very long time to run and I expect their size to be many Gbs. 

Thank you again for your help

**Comment by pz-white** [comment]
Hi @SamuelMidgley 
Could you upload the similarity file to github as I cannot open the drive link, thanks.

**Comment by SamuelMidgley** [comment]
The file is too large to upload to github, i have uploaded and turned on link sharing restrictions so should work now [link](https://drive.google.com/file/d/1ojLN4sX8au2EQzHfptpR5NJxY_UyxU0s/view?usp=sharing)

**Comment by SamuelMidgley** [comment]
let me know if there is still a problem with this though @pz-white

**Comment by SamuelMidgley** [comment]
Hi @pz-white 
After running the code including the drug similarity matrix with lr=0.005 it helps and gets a lot lower but still doesn't reach the same point as the Base Matrix Factorisation, I'm going to move it to HPC and try with a lower learning rate and see if that improves it further. 
![image](https://user-images.githubusercontent.com/80775550/118015929-b6c23780-b34c-11eb-99e1-5189b9897b6e.png)

With regards to the target similarity matrix, it takes roughly 80 minutes to find 5 full rows of the similarity matrix which at that rate would take about 11 days to complete so has some serious performance issues. Am I missing a library that has an optimised normalised smith-waterman score?

**Comment by pz-white** [comment]
Hi @SamuelMidgley 
- About the target similarity, although you replaced it with a learned matrix but it is not appropriate. The MSCMF architecture is changed by the operation and the performance may not be the expected. I suggest you should still have a target similarity matrix instead directly dropping it. Here I found a tool [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) for generating the protein similarity and you can leverge it since I see your own target similarity calculation is very time-consuming.
- For your drug_sim file, the size is very large as you saved a high precision float. I suggest keeping only **four decimal places** (same as to target_sim later) to speed up training time.
- The 600 epoches are too many and please limit to less than 300 after introducing protein similarity matrix.
Thanks.

**Comment by pz-white** [comment]
@SamuelMidgley 
As I'm all available today, you can directly find me on Skype you need to further discuss (no matter message or video). But please note that I'm still in China, we have 7 hours of time difference. It is better to call me before 3pm of UK local time. Thanks.

**Comment by SamuelMidgley** [comment]
Hi @pz-white
- The learned matrix is not the substitute for the target matrix it was a new thing I added that I do not think is correct.  What is the best method for including the similarity matrix in the forward pass? Do I have to calculate `AA^T` for the whole of A and then compare this to the full similarity matrix for drugs. Or somehow change the dimensions to accommodate the batch? Thank you for the link to BLAST I will use this to make the target similarity matrix.
- Okay I will do that
- Is there anything wrong with training a model for 600 epochs or is it bad practice representative of a poor model?

Thank you very much, I will first try the things you have suggested but I will probably need a call later on, if needed this will hopefully be done around 12pm UK time or whenever is convenient for you around that time.
